Use generated serialization for NSData

sheeparegreat opened the pull request:
#### 0ee026d9d1d0e2a92dd329244493d8742ccef070
<pre>
Use generated serialization for NSData
<a href="https://bugs.webkit.org/show_bug.cgi?id=263378">https://bugs.webkit.org/show_bug.cgi?id=263378</a>
rdar://117206485

Reviewed by Brady Eidson and Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeDataInternal):
(IPC::decodeDataInternal):
* Source/WebKit/Shared/Cocoa/CoreIPCData.h: Added.
(WebKit::CoreIPCData::CoreIPCData):
(WebKit::CoreIPCData::createData const):
(WebKit::CoreIPCData::get const):
* Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/269766@main">https://commits.webkit.org/269766@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c31c327fe48ee474fb8ecb76b87d399f6ff706ee

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23387 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1501 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24510 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21605 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23940 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23628 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1064 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26151 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/870 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27289 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21330 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25160 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/856 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18596 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/813 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5612 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1268 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1117 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+    RetainPtr<CFDataRef> takeData()
+    {
+        return WTFMove(m_data);
+    };
```
unneeded semicolon.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+    {
+    }
+
+    CoreIPCData(const IPC::DataReference& d)
```
Why is this needed?  Seems suspicious.  It looks like unneeded data copies.

sheeparegreat replied:
Because the decoder will make an IPC::DataReference and call into this constructor

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+
+#import "DataReference.h"
+
+#import <Foundation/Foundation.h>
```
This probably shouldn't need Foundation, just CoreFoundation.h

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+
+#import <Foundation/Foundation.h>
+#import <cstdint>
+#import <span>
```
This is redundant with DataReference.h

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCData, void>;
+    RetainPtr<CFDataRef> m_data;
+    IPC::DataReference m_ref;
```
This actually doesn't need to be stored.  Just make a function like this:
IPC::DataReference asDataReference() { ... }

Then make a constructor that takes a span/DataReference.  Then you only serialize the result of calling this function, and then you don't need to use NotSerialized, which is usually an indication that something needs to be refactored or fixed.

sheeparegreat replied:
A single member now which is an IPC::DataReference

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+
+class CoreIPCData {
+public:
+    CoreIPCData();
```
Remove.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,68 @@
+public:
+    CoreIPCData();
+
+    CoreIPCData(CFDataRef d)
```
This constructor needs to set the RetainPtr.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,56 @@
+    }
+
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCData, void>;
```
I think it would be even nicer if you made an accessor.  This isn't a great reason to use "friend"

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,56 @@
+    {
+    }
+
+    RetainPtr<CFDataRef> createData()
```
const

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,56 @@
+
+    RetainPtr<CFDataRef> createData()
+    {
+        return { adoptCF(CFDataCreate(0, m_ref.data(), m_ref.size())) };
```
{ and } unnecessary

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,56 @@
+
+private:
+    friend struct IPC::ArgumentCoder<CoreIPCData, void>;
+    IPC::DataReference m_ref;
```
We usually try to avoid abbreviations in our member names.
m_reference
I think it might be even better to just use a std::span (which is interchangeable with DataReference) and call this m_span

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCData.h:
```diff
@@ -0,0 +1,56 @@
+
+class CoreIPCData {
+public:
+    CoreIPCData(const IPC::DataReference& d)
```
d -> data
More meaningful parameter names.

sheeparegreat committed:
Use generated serialization for NSData
https://bugs.webkit.org/show_bug.cgi?id=263378
rdar://117206485

Reviewed by Brady Eidson and Alex Christensen.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeDataInternal):
(IPC::decodeDataInternal):
* Source/WebKit/Shared/Cocoa/CoreIPCData.h: Added.
(WebKit::CoreIPCData::CoreIPCData):
(WebKit::CoreIPCData::createData const):
(WebKit::CoreIPCData::get const):
* Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/269766@main
